### PR TITLE
Report ARIA details role

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:

None

### Summary of the issue:

When announcing if an object has details, more information can be provided about the role of the details.
For example, if a comment is tied to a word, "has comment" should be reported instead of "has details".

### Description of how this pull request fixes the issue:
Collects the details role.

The attribute details role is determined in a number of cases.

#### Braille
Use `has cmmt` to report presence of comment role aria-details target. [Discussion](https://github.com/nvaccess/nvda/pull/13649#discussion_r901200628)

#### Focus Mode

detailsRole is normalized on the NVDAObject level to a `controlType.Role`.

##### Chromium
The IA2 attribute `details-roles` is used to get the role.

This supports the following roles ([code](https://source.chromium.org/chromium/chromium/src/+/main:ui/accessibility/platform/ax_platform_node_base.cc;l=2337;drc=439c6ac35199f3828faf0f0d070d84109c9bf6a1;bpv=0;bpt=1)): `comment, definition, doc-endnote, doc-footnote`. Other roles are given a value of * .

Currently only chromium supports `details-roles`.
Additionally,  `doc-endnote` and `definition` are not supported IA2 attributes, and are not supported in NVDA widely.

We limit reporting for all cases by only reporting the fully supported `details-roles`: `doc-footnote, comment`

Other roles are reported as "has details".

##### FireFox

The related details node is fetched using in process injection, and the IAccessible information is retrieved.
The role is determined from the IAccessible role of the related node.

This is slower and provides a wider reporting of roles.

#### Browse Mode

The detailsRole attribute is added directly to the buffer, either as a descriptive role string or a role integer as a string.

### Testing strategy:

The system test cases test the details role for Chrome, using both focus and browse mode.

`runsystemtests.bat -t "*aria details*"`

Equivalent manual testing was performed using those test cases on both Chrome and FireFox.

### Known issues with pull request:
Braille tests will not work as expected as braille only reports details in marked content. #13815

### Change log entries:
New features
```t2t
- Instead of reporting "has details" the purpose of details is included where possible, for example "has comment".
```

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

File: source/braille.py [hunk 1]
```diff
@@ -546,7 +551,14 @@ def getPropertiesBraille(**propertyValues) -> str:  # noqa: C901
 		textList.append(description)
 	hasDetails = propertyValues.get("hasDetails")
 	if hasDetails:
-		textList.append("details")
+		detailsRole = propertyValues.get("detailsRole", controlTypes.Role.UNKNOWN)
+		if detailsRole != controlTypes.Role.UNKNOWN:
+			textList.append(roleLabels.get(detailsRole, detailsRole.displayString))
```
Comment (seanbudd) [inline source/braille.py:556]: Should this be something like "has cmnt", "has sggstn", etc instead?

Comment (feerrenrut) [reply source/braille.py:558]: Do those roles have a shorthand labels already? You can potentially combine them programmatically.

Comment (seanbudd) [reply source/braille.py]: Yes they do - the current behaviour is to just report "cmnt", "fnote", "details" as opposed to "has details"/"has fnote"

File: source/aria.py [hunk 2]
```diff
@@ -1,13 +1,60 @@
...
 
-from typing import Dict
+from typing import Dict, Union
 from enum import Enum
 import controlTypes
+from logHandler import log
 
-ariaRolesToNVDARoles: Dict[str, int] = {
+
+def normalizeDetailsRole(detailsRole: Union[str, controlTypes.Role, None]) -> controlTypes.Role:
```
Comment (feerrenrut) [inline source/aria.py:12]: I'm not sure this belongs here either.
Also, I'm not sure about the union approach. The code sharing is minimal. The downside is thinking about two very different code paths.

Are there other examples like this?

File: source/NVDAObjects/IAccessible/mozilla.py [hunk 3]
```diff
@@ -73,6 +74,17 @@ def _get_detailsSummary(self) -> Optional[str]:
 			# just take the first for now.
 			return target.summarizeInProcess()
 
+	def _get_detailsRole(self) -> Optional[controlTypes.Role]:
+		# Unlike base Ia2Web implementation, the details-roles
+		# IA2 attribute is not exposed in FireFox.
+		# Although slower, we have to fetch the details relations instead.
+		detailsRelations = self.detailsRelations
+		if not detailsRelations:
+			return None
```
Comment (feerrenrut) [inline source/NVDAObjects/IAccessible/mozilla.py:83]: This relies on `detailsRelations` observing `config.conf["debugLog"]["annotations"]`. Easy for that to change and break this.

File: source/NVDAObjects/IAccessible/ia2Web.py [hunk 4]
```diff
@@ -62,6 +62,18 @@ def _get_detailsSummary(self) -> typing.Optional[str]:
...
 		return bool(self.IA2Attributes.get("details-roles"))
 
+	def _get_detailsRole(self) -> typing.Optional[controlTypes.Role]:
+		# Currently only defined in Chrome as of May 2022
+		# Refer to ComputeDetailsRoles
+		# https://chromium.googlesource.com/chromium/src/+/main/ui/accessibility/platform/ax_platform_node_base.cc#2419
+		detailsRoles = self.IA2Attributes.get("details-roles")
+		if not detailsRoles:
+			if config.conf["debugLog"]["annotations"]:
+				log.debug("no details-roles")
```
Comment (feerrenrut) [inline source/NVDAObjects/IAccessible/ia2Web.py:72]: This is a misleading log message.

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 5]
```diff
@@ -361,9 +361,61 @@ CComPtr<IAccessible2> getTextBoxInComboBox(
...
+		}
+		if (varRole.vt == VT_I4) {
+			*role = varRole.lVal;
+		}
+		else if (varRole.vt == VT_BSTR) {
+			*roleString = varRole.bstrVal;
+		}
+	}
+	VariantClear(&varRole);
+}
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp:383]: This should be using RAII types for `pacc, varChild, roleString, varRole`. 
See `appendNameDescription` in `nvdaHelper/remote/textFromIAccessible.cpp` for examples.

Rather than taking pointers, return a `std::tuple` to make use of auto unpacking https://en.cppreference.com/w/cpp/language/structured_binding

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.h [hunk 6]
```diff
@@ -33,6 +33,14 @@ class GeckoVBufBackend_t: public VBufBackend_t {
 		bool ignoreInteractiveUnlabelledGraphics=false
 	);
 
+	void fillVBufAriaDetails(
+		int docHandle,
+		IAccessible2* pacc,
+		VBufStorage_buffer_t* buffer,
+		VBufStorage_controlFieldNode_t* parentNode,
+		std::wstring roleAttr
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.h:41]: this will copy the string, take by reference `std::wstring& roleAttr`

File: nvdaHelper/vbufBase/storage.cpp [hunk 7]
```diff
@@ -321,6 +321,17 @@ bool VBufStorage_fieldNode_t::addAttribute(const std::wstring& name, const std::
 	return true;
 }
 
+bool VBufStorage_fieldNode_t::getAttribute(const std::wstring& name, std::wstring* value) {
```
Comment (feerrenrut) [inline nvdaHelper/vbufBase/storage.cpp:324]: Don't take `std::wstring* value` as a pointer. Taking by reference would be more appropriate as it implies can't be `nullptr`. Better would be to return a `std::optional`.

File: nvdaHelper/vbufBase/storage.cpp [hunk 8]
```diff
@@ -321,6 +321,17 @@ bool VBufStorage_fieldNode_t::addAttribute(const std::wstring& name, const std::
 	return true;
 }
 
+bool VBufStorage_fieldNode_t::getAttribute(const std::wstring& name, std::wstring* value) {
+	LOG_DEBUG(L"Getting attribute " << name);
+	VBufStorage_attributeMap_t::const_iterator foundAttrib = attributes.find(name);
```
Comment (feerrenrut) [inline nvdaHelper/vbufBase/storage.cpp:326]: You can just do `auto foundItr = attributes.find(name);`
If you mark the function as `const`, then `attributes` will be `const` and return a `const` iterator.

File: nvdaHelper/vbufBase/storage.h [hunk 9]
```diff
@@ -266,6 +266,14 @@ class VBufStorage_fieldNode_t {
  */
 	bool addAttribute(const std::wstring& name, const std::wstring& value);
 
+/**
+ * Gets an attribute value for this field.
+ * @param name the name of the attribute
+ * @param value where to store the value for the attribute.
+ * @return true if the attribute exists, false if it doesn't exist.
+ */
+	bool getAttribute(const std::wstring& name, std::wstring* value);
```
Comment (feerrenrut) [inline nvdaHelper/vbufBase/storage.h:275]: I don't know why all these comments don't match the identation, but it is unusual. I would assume there is no reason for it. Double check that doxygen generates docs and I believe it is fine

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/aria.py (hunk 2 above)
Comment (seanbudd) [reply source/aria.py:12]: Split for focus and browse mode

File: source/braille.py (hunk 1 above)
Comment (seanbudd) [reply source/braille.py]: @michaelDCurran / @leonardder / any other braille users.
When navigating an object with details, the current behaviour is to report "details" for braille and "has details" for speech. 
When navigating an object with a known type of details, the behaviour for speech is to report "has comment".

Reporting "cmnt" for braille is misleading. Is "has cmnt" appropriate for braille? Is there a shorter alternative?

File: tests/system/robot/chromeTests.py [hunk 10]
```diff
@@ -140,6 +140,125 @@ def test_mark_aria_details_FreeReviewCursor():
...
+				<span aria-details="definition-details">definition,</span>
+				<span aria-details="unknown-details">form</span>
+			</p>
+		</div>
+		<div>
+			<p>
+				<div id="endnote-details" role="doc-endnote">details with role doc-endnote</div>
+				<div id="footnote-details" role="doc-footnote">details with role doc-footnote</div>
+				<div id="comment-details" role="comment">details with role comment</div>
+				<div id="definition-details">details with role definition</div>
```
Comment (feerrenrut) [inline tests/system/robot/chromeTests.py:160]: This doesn't actually have a role, further it doesn't seem to match examples for how this should work.
See:
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dfn
- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/definition_role

It would be good to check the other items too.

File: tests/system/robot/chromeTests.py [hunk 11]
```diff
@@ -140,6 +140,125 @@ def test_mark_aria_details_FreeReviewCursor():
...
+				<div id="unknown-details" role="form">details with role form</div>
```
Comment (feerrenrut) [inline tests/system/robot/chromeTests.py:161]: Why is this here? Because it is unsupported, please add a comment in the test data to clarify.

File: tests/system/robot/chromeTests.py (hunk 10 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py:160]: yes this was a mistake when submitting code. Locally, I have tested both `<definition>` and `<definition role="definition">`. Both of these report the wrong role, both on the accessibility tree and in NVDA.

File: tests/system/robot/chromeTests.py (hunk 11 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py:161]: Yes, to test an unsupported role

File: source/NVDAObjects/IAccessible/mozilla.py (hunk 3 above)
Comment (seanbudd) [reply source/NVDAObjects/IAccessible/mozilla.py:83]: this is for debug logging.

`config.conf["annotations"]["reportDetails"]` is checked in `speech.py` and `braille.py`

File: source/NVDAObjects/IAccessible/ia2Web.py (hunk 4 above)
Comment (seanbudd) [reply source/NVDAObjects/IAccessible/ia2Web.py:72]: how so? I've updated it to "details-roles not found"

File: tests/system/robot/chromeTests.py (hunk 11 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py:161]: This is documented in comments in `expectedSpeech` and `expectedBraille`.

I'm uncertain on the best way to add comments to the html multi-line string

File: source/aria.py (hunk 2 above)
Comment (seanbudd) [reply source/aria.py:12]: This function has been removed in favour of handling the 3 cases separately (Firefox focus mode, Chrome Focus mode, Browse mode).

File: tests/system/robot/chromeTests.py [hunk 12]
```diff
@@ -140,6 +140,125 @@ def test_mark_aria_details_FreeReviewCursor():
...
+		"""
+	)
+	expectedSpeech = SPEECH_SEP.join([
+		"edit",
+		"multi line",
+		# the role doc-endnote is unsupported as an IA2 role
+		# The role "list item" is used instead
+		"has details",
+		"doc endnote,",
+		"",
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply tests/system/robot/chromeTests.py:173]: I believe these come from navigating between spans

File: source/braille.py [hunk 13]
```diff
@@ -551,9 +553,11 @@ def getPropertiesBraille(**propertyValues) -> str:  # noqa: C901
 		textList.append(description)
 	hasDetails = propertyValues.get("hasDetails")
 	if hasDetails:
-		detailsRole = propertyValues.get("detailsRole", controlTypes.Role.UNKNOWN)
+		detailsRole = aria.normalizeDetailsRole(propertyValues.get("detailsRole"))
 		if detailsRole != controlTypes.Role.UNKNOWN:
-			textList.append(roleLabels.get(detailsRole, detailsRole.displayString))
+			detailsRoleLabel = roleLabels.get(detailsRole, detailsRole.displayString)
+			# TODO: should this be `_("has %s") % detailsRoleLabel`?
+			textList.append(detailsRoleLabel)
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/braille.py:560]: That has been changed, with no replies to the earlier conversation: https://github.com/nvaccess/nvda/pull/13649#discussion_r879979596

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp (hunk 5 above)
Comment (seanbudd) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: I can't figure out how to do this for the `VARIANT` type `varChild`. Same with `VBufStorage_buffer_t* buffer, VBufStorage_controlFieldNode_t* parentNode` in https://github.com/nvaccess/nvda/pull/13649#discussion_r874758216

File: nvdaHelper/vbufBase/storage.h (hunk 9 above)
Comment (seanbudd) [reply nvdaHelper/vbufBase/storage.h]: Running DOxygen from NVDAHelper doesn't appear to build anything from docstrings of functions. I'm not sure what the expected output of DOxygen is, but it appears to just be generic notes written about *.dlls

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.h [hunk 14]
```diff
@@ -33,6 +33,14 @@ class GeckoVBufBackend_t: public VBufBackend_t {
 		bool ignoreInteractiveUnlabelledGraphics=false
 	);
 
+	void fillVBufAriaDetails(
+		int docHandle,
+		CComPtr<IAccessible2> pacc,
+		VBufStorage_buffer_t* buffer,
+		VBufStorage_controlFieldNode_t* parentNode,
+		std::wstring& roleAttr
+	);
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.h:42]: If this isn't called by any other class, and doesn't use any member variables or functions, it can doesn't need to be declared in the header. It can be entirely within the `cpp` file.

File: source/NVDAObjects/IAccessible/mozilla.py (hunk 3 above)
Comment (feerrenrut) [reply source/NVDAObjects/IAccessible/mozilla.py:83]: ah yes, I must have misread that.

File: source/NVDAObjects/IAccessible/mozilla.py [hunk 15]
```diff
@@ -83,7 +83,14 @@ def _get_detailsRole(self) -> Optional[controlTypes.Role]:
 			return None
 		for target in detailsRelations:
 			# just take the first for now.
-			return aria.normalizeDetailsRole(target.IAccessibleRole)
+			from .chromium import supportedAriaDetailsRoles
```
Comment (feerrenrut) [inline source/NVDAObjects/IAccessible/mozilla.py:86]: Perhaps a note here that firefox may diverge in the future, and these values are just used out of convenience.

File: source/braille.py [hunk 16]
```diff
@@ -553,11 +552,10 @@ def getPropertiesBraille(**propertyValues) -> str:  # noqa: C901
...
 	hasDetails = propertyValues.get("hasDetails")
 	if hasDetails:
-		detailsRole = aria.normalizeDetailsRole(propertyValues.get("detailsRole"))
-		if detailsRole != controlTypes.Role.UNKNOWN:
+		detailsRole: Optional[controlTypes.Role] = propertyValues.get("detailsRole")
+		if detailsRole is not None:
 			detailsRoleLabel = roleLabels.get(detailsRole, detailsRole.displayString)
-			# TODO: should this be `_("has %s") % detailsRoleLabel`?
-			textList.append(detailsRoleLabel)
+			textList.append(_("has %s") % detailsRoleLabel)
```
Comment (feerrenrut) [inline source/braille.py:558]: This needs a translation comment.

File: tests/system/robot/chromeTests.py (hunk 11 above)
Comment (feerrenrut) [reply tests/system/robot/chromeTests.py:163]: You could:
- add a html comment `<!-- a note here -->`
- Split the string and add a python comment in between

``` python
"""
some string contents
""" 
# the next line is here for xyz
"""
some interesting line
"""
```

I think the html comment is the better option, then it is in the test case when loaded by the browser.

File: tests/system/robot/chromeTests.py [hunk 17]
```diff
@@ -220,7 +231,7 @@ def test_mark_aria_details_role():
 		actualBraille,
 		# TODO: fix known bug with braille not announcing details
 		# expectedBraille,
```
Comment (feerrenrut) [inline tests/system/robot/chromeTests.py:233]: What is the plan to deal with this?

I note it is listed in known issues, can you make sure that a bug is raised/linked for it please.

File: tests/system/robot/chromeTests.py [hunk 18]
```diff
@@ -157,7 +158,8 @@ def test_mark_aria_details_role():
 				<div id="endnote-details" role="doc-endnote">details with role doc-endnote</div>
 				<div id="footnote-details" role="doc-footnote">details with role doc-footnote</div>
 				<div id="comment-details" role="comment">details with role comment</div>
-				<div id="definition-details">details with role definition</div>
+				<div id="definition-details-as-role" role="definition">details with role definition</div>
+				<definition id="definition-details-as-tag">details with tag definition</definition>
```
Comment (feerrenrut) [inline tests/system/robot/chromeTests.py:162]: I don't think this is right. See:
- https://codepen.io/reefturner/pen/PoQXKeJ
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dfn
- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/definition_role

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: nvdaHelper/vbufBase/storage.h (hunk 9 above)
Comment (feerrenrut) [reply nvdaHelper/vbufBase/storage.h]: Hmm that doesn't sound right. Let's investigate this tomorrow.

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp (hunk 5 above)
Comment (feerrenrut) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: We can talk about this tomorrow.

File: tests/system/robot/chromeTests.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (feerrenrut) [reply tests/system/robot/chromeTests.py]: Can you add a comment to explain what causes them.

[the remark this answers is not in the document]
Comment (seanbudd) [reply tests/system/robot/chromeTests.py]: that has already been done

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 19]
```diff
@@ -362,60 +362,65 @@ CComPtr<IAccessible2> getTextBoxInComboBox(
...
 		}
 	}
 	if (detailsForId.has_value()) {
 		detailsParentNode = buffer->getControlFieldNodeWithIdentifier(docHandle, detailsForId.value());
-		detailsRole.assign(roleAttr);
+		detailsRole = roleAttr;
 	}
-	if (detailsParentNode != nullptr && !detailsRole.empty()) {
-		detailsParentNode->addAttribute(L"detailsRole", detailsRole);
+	if (detailsParentNode != nullptr && detailsRole.has_value() && !detailsRole.value().empty()) {
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp:418]: I think that should be fine.

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.h (hunk 14 above)
Comment (seanbudd) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.h:42]: `fillVBufAriaDetails` uses the member function `getRelationId`

File: tests/system/robot/chromeTests.py (hunk 17 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py]: I am yet to write up an issue for this.

File: nvdaHelper/vbufBase/storage.h (hunk 9 above)
Comment (seanbudd) [reply nvdaHelper/vbufBase/storage.h]: Thanks, I've added a reference to Doxygen in the relevant section of the readme.
It wasn't clear that I was looking to run `scons devDocs_nvdaHelper` based on the reference to Doxygen in the initial review comment.

Comment (seanbudd) [reply nvdaHelper/vbufBase/storage.h]: I have fixed the indentation on this PR, but the other items (with different indentation) also show up as expected in the generated documentation.

File: tests/system/robot/chromeTests.py [hunk 20]
```diff
@@ -140,6 +140,138 @@ def test_mark_aria_details_FreeReviewCursor():
...
+				<span aria-details="unknown-details">form</span>
+			</p>
+		</div>
+		<div>
+			<p>
+				<!-- Supported by Chrome attribute details-roles -->
+				<div id="endnote-details" role="doc-endnote">details with role doc-endnote</div>
+				<div id="footnote-details" role="doc-footnote">details with role doc-footnote</div>
+				<div id="comment-details" role="comment">details with role comment</div>
+				<div id="definition-details-as-role" role="definition">details with role definition</div>
```
Comment (feerrenrut) [inline tests/system/robot/chromeTests.py:162]: I think this needs a `role="term"` next to it for it to be valid.

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 21]
```diff
@@ -361,9 +361,64 @@ CComPtr<IAccessible2> getTextBoxInComboBox(
...
+	CComPtr<IAccessible2> pacc,
+	VBufStorage_buffer_t& buffer,
+	VBufStorage_controlFieldNode_t& parentNode,
+	const std::wstring& roleAttr
+){
+	/* Set the details role by checking for both IA2_RELATION_DETAILS and IA2_RELATION_DETAILS_FOR as one
+	of the nodes in the relationship will not be in the buffer yet */
+	std::optional<int> detailsId = getRelationId(IA2_RELATION_DETAILS, pacc);
+	std::optional<int> detailsForId = getRelationId(IA2_RELATION_DETAILS_FOR, pacc);
+	VBufStorage_controlFieldNode_t* detailsParentNode;
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp:402]: Don't declare a pointer (or any other POD type) without initializing it. 
>[Read from an indeterminate byte](https://en.cppreference.com/w/cpp/language/default_initialization)
Use of an indeterminate value obtained by default-initializing a non-class variable of any type is [undefined behavior](https://en.cppreference.com/w/cpp/language/ub) (in particular, it may be a [trap representation](https://en.cppreference.com/w/cpp/language/object#Object_representation_and_value_representation)), except in the following cases....

Practically, it may hold a random non zero (therefore not `nullptr`) value, while still being invalid. Causing an error around line 416.

```suggestion
	VBufStorage_controlFieldNode_t* detailsParentNode = nullptr;

```

Comment (feerrenrut) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp:402]: This is picked when you build with `scons source nvdaHelperDebugFlags=analyze`
>`...x86\vbufBackends\gecko_ia2\gecko_ia2.cpp(416) : warning C6001: Using uninitialized memory 'detailsParentNode'.: Lines: 400, 401, 402, 403, 404, 412, 416`

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 22]
```diff
@@ -436,21 +492,11 @@ VBufStorage_fieldNode_t* GeckoVBufBackend_t::fillVBuf(
...
-		if(varRole.vt==VT_I4)
-			role=varRole.lVal;
-		else if(varRole.vt==VT_BSTR)
-			roleString=varRole.bstrVal;
-	}
+	long role = 0;
+	CComBSTR roleString;
+	CComPtr<IAccessible2> smartPacc = CComQIPtr<IAccessible2>(pacc);
+	CComVariant smartVariant = CComVariant(&varChild);
+	std::tie(role, roleString) = getRoleLongRoleString(smartPacc, smartVariant);
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp:499]: Probably a good idea to limit the scope here. The scope of `smartPacc` and `smartVariant` defines when they will do their clean up. You can add arbitrary `{}` characters to limit scope, though in this case you might consider just passing in the two parameters directly.

 ```suggestion
	std::tie(role, roleString) = getRoleLongRoleString(CComQIPtr<IAccessible2>(pacc), CComVariant(&varChild));

```

File: source/NVDAObjects/IAccessible/mozilla.py (hunk 15 above)
Comment (feerrenrut) [reply source/NVDAObjects/IAccessible/mozilla.py:86]: I think this was missed?

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/controlTypes/role.py [hunk 23]
```diff
@@ -177,6 +177,9 @@ def _displayStringLabels(self):
 	MARKED_CONTENT = 153
 	BUSY_INDICATOR = 154  # Used for progress bars with indeterminate state
 	# To maintain backwards compatibility, above Roles must maintain their values.
+	COMMENT = auto()
```
Comment (lukaszgo1) [inline source/controlTypes/role.py:180]: What exact backwards compatibility guarantees we offer add-on developers with regard to numeric values of roles? If we guarantee only that all roles prior to 2022.1 maintain their integer values then using `auto` here is fine. On the other hand since roles are not numbered sequentially I can imagine a case in which the newly added roles are assigned a value which is currently not taken but is in the middle of other roles, and some add-on relies on the missing integers for example by checking roles with less than or greater than operator.

File: source/NVDAObjects/IAccessible/mozilla.py (hunk 15 above)
Comment (seanbudd) [reply source/NVDAObjects/IAccessible/mozilla.py:88]: yep - looks like I forgot to commit the change. Fixed now.

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp (hunk 22 above)
Comment (seanbudd) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp:499]: `smartPacc` is used further down in the same scope by `fillVBufAriaDetails`. I can make this change for `smartVariant`

File: source/controlTypes/role.py (hunk 23 above)
Comment (seanbudd) [reply source/controlTypes/role.py:180]: >What exact backwards compatibility guarantees we offer add-on developers with regard to numeric values of roles?

Given the changes that are already in 2022.1, I don't think we can make any guarantees in regards to numeric values of roles. The changes in 2022.1 dropped support for numeric values of roles, but kept aliases for ease of use. 

>some add-on relies on the missing integers for example by checking roles with less than or greater than operator

This is unsupported. Add-on authors should not be relying on this to begin with. Roles could be created at any point which used "missing integers". Using "missing integers" would also cause conflict with other add-ons. Using Enums discourages these approaches by design. The NVDA API doesn't have "reserved space" for add-ons as it stands.

File: tests/system/robot/chromeTests.py (hunk 18 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py]: I have removed the use of `<dfn>`.
`<dfn>` is used for the term of the definition, and maps to the text of the parent node.
I don't believe this syntax is useful / supported by `aria-details`.

```html
<p><dfn>Mansplaining</dfn>, a portmanteau of "man" and "explain", is the patronizing act of explaining without being asked to do so, to someone already learned on the topic, often after someone has already explained it. </p>
```

File: source/braille.py (hunk 1 above)
Comment (seanbudd) [reply source/braille.py]: No - but in absence of a clear answer, "has cmnt" has been suggested by @michaelDCurran to be the best option.

File: tests/system/robot/chromeTests.py (hunk 17 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py]: Issue created: #13815

File: source/controlTypes/role.py (hunk 23 above)
Comment (seanbudd) [reply source/controlTypes/role.py:180]: @feerrenrut - thoughts?

File: tests/system/robot/chromeTests.py (hunk 18 above)
Comment (feerrenrut) [reply tests/system/robot/chromeTests.py]: MDN suggests that the `dfn` elements should be preferred over the alternative approach. Even if unsupported, I think it worth having in the test.

Web devs will likely expect to be able to the following and get "has definition". Although still unsupported, recording the use-case for future consideration will help to prevent missing it later.
```HTML
<div>
    <p>Some topic that may be related to <span aria-details="def1">mansplaining</span>.</p>
</div>
<div>
    <p><dfn id="def1">Mansplaining</dfn>, a portmanteau of "man" and "explain", is the patronizing act of explaining</p>
</div>
```

That said, I agree this looks difficult to support, there seems to be a disconnect between term vs definition for the term.

Comment (seanbudd) [reply tests/system/robot/chromeTests.py]: I don't believe this syntax would be appropriate, as the developer would be intending to target the definition of the term, which is the unlabelled `<p>` tag.

File: source/controlTypes/role.py (hunk 23 above)
Comment (seanbudd) [reply source/controlTypes/role.py]: @lukaszgo1 - we are going to err on the side of caution here in trying to minimize introducing a breaking change. I've added some notes to clarify the position in bf4e4b2.

As per the notes and earlier comment, it is difficult to guarantee full support for treating `Role`s as integers.

File: tests/system/robot/chromeTests.py (hunk 20 above)
Comment (feerrenrut) [reply tests/system/robot/chromeTests.py:161]: I still don't think this is quite right. [Shouldn't it be](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/definition_role):
```suggestion
				<p><span role="term">definition</span>:	<span id="definition-details-as-role" role="definition">details with role definition</span>
				</p>
```

File: source/controlTypes/role.py (hunk 23 above)
Comment (feerrenrut) [reply source/controlTypes/role.py]: We've discussed this, @seanbudd will revert the use of `enum.auto` and add some comments to the Role enum to serve as future reference for the decisions we made around this.

File: source/controlTypes/role.py [hunk 24]
```diff
@@ -14,15 +13,36 @@
...
+	From a string, this can be done as `controlTypes.Role[nameOfRole]` e.g. `Role["CHECKBOX"]`.
+	Although unlikely to change, if names/values changing represents a significant risk for your add-on,
+	then consider decoupling, and maintain an internal mapping of `Role` to add-on internal roles.
+
+	As add-on authors may still rely on the values, new members of `Role` should continue
+	the previous pattern of incrementing.
+	Using `enum.auto` may backfill role values. For example, creating a new role with auto value
+	68 would cause it to be included in an add-on which checks roles as follows:
+	`(66 < role) and (role < 70)`.
+
```
Comment (feerrenrut) [inline source/controlTypes/role.py:27]: Behavior that is supported, and intentionally persisted:
```python
r = ROLE_GRAPHIC  # intentionally alias use

assert isinstance(r, int)
assert r == 16
assert 15 < r < 17
```

Perhaps these should be added as unit tests and the comment point to them?

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/controlTypes/role.py (hunk 24 above)
Comment (seanbudd) [reply source/controlTypes/role.py:26]: That has been done in 0ad0d35

File: tests/system/robot/chromeTests.py (hunk 20 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py:161]: Fixed by 55eeb9f

File: tests/system/robot/chromeTests.py [hunk 25]
```diff
@@ -159,6 +159,10 @@ def test_mark_aria_details_role():
 				<div id="footnote-details" role="doc-footnote">details with role doc-footnote</div>
 				<div id="comment-details" role="comment">details with role comment</div>
 				<div id="definition-details-as-role" role="definition">details with role definition</div>
+				<p>
+					<span role="term">definition</span>:
+					<span id="definition-details-as-role" role="definition">details with role definition</span>
+				</p>
```
Comment (feerrenrut) [inline tests/system/robot/chromeTests.py:165]: These shouldn't both exist (two elements with `id="definition-details-as-role"`.
I'd also like the `<dfn>` example added. It seems both the `role="term"` / `role="definition"` case and the `<p> A description for a <dfn>term</dfn> is in this paragraph.</p>` case share the similar problem about `aria-details` pointing to the **term** rather than the **definition**. Illustrating this in the tests will be helpful when having the discussion later.

File: source/braille.py (hunk 1 above)
Comment (feerrenrut) [reply source/braille.py]: I'm editing the description to describe the outcome and point to this conversation.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: tests/system/robot/chromeTests.py (hunk 25 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py]: I've added documentation to the test samples and fixed it so we consistently point to the term in cfb74bf

**Review by feerrenrut — APPROVED** [review]

**Comment by feerrenrut** [comment]
Although the test failures look related to this change, similar tests have failed on other recent PR builds. I have started the build again.
